pkg/controller/storageversionmigrator/storageversionmigrator.go

```diff
@@ -0,0 +1,294 @@
+	}
+
+	klog.FromContext(ctx).V(2).Info("Error syncing SVM resource, retrying", "svm", key, "err", err)
+	svmc.queue.AddRateLimited(key)
+
+	return true
+}
+
+func (svmc *SVMController) sync(ctx context.Context, key string) error {
+	logger := klog.FromContext(ctx)
+	startTime := time.Now()
+
+	// SVM is a cluster scoped resource so we don't care about the namespace
+	_, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	svm, err := svmc.svmListers.Get(name)
+	if err != nil {
+		return err
+	}
+	// working with a copy to avoid race condition between this and resource version controller
+	toBeProcessedSVM := svm.DeepCopy()
+
+	if HasCondition(toBeProcessedSVM, svmv1alpha1.MigrationSucceeded) || HasCondition(toBeProcessedSVM, svmv1alpha1.MigrationFailed) {
+		logger.V(4).Info("Migration has already succeeded or failed previously, skipping", "svm", name)
+		return nil
+	}
+
+	if toBeProcessedSVM.Status.ResourceVersion == "" {
+		logger.V(4).Info("list resource version is empty, requeuing to attempt again", "svm", name)
+		svmc.queue.AddRateLimited(key)
+		return nil
+	}
+	gvr := getGVRFromResource(toBeProcessedSVM)
+
+	// wait for the dependency graph to be synchronized before migration to avoid a nil pointer.
+	// this prevents deadlocks in scenarios where available resources have changed, and one of the informer caches will never fill.
+	if !cache.WaitForNamedCacheSync(svmc.gcCollector.Name(), ctx.Done(), func() bool {
+		return svmc.gcCollector.IsSynced(logger)
+	}) {
+		return fmt.Errorf("dependency graph is not synced, requeuing to attempt again")
+	}
+
+	gcListResourceVersion, err := convertResourceVersionToInt(svmc.gcCollector.GetMonitorController(gvr).LastSyncResourceVersion())
+	if err != nil {
+		return err
+	}
+	listResourceVersion, err := convertResourceVersionToInt(toBeProcessedSVM.Status.ResourceVersion)
+	if err != nil {
+		return err
+	}
+
+	if gcListResourceVersion < listResourceVersion {
+		return fmt.Errorf("GC cache is not up to date, requeuing to attempt again. gcListResourceVersion: %d, listResourceVersion: %d", gcListResourceVersion, listResourceVersion)
+	}
+
+	toBeProcessedSVM, err = svmc.kubeClient.StoragemigrationV1alpha1().
+		StorageVersionMigrations().
+		UpdateStatus(
+			ctx,
+			setStatusConditions(toBeProcessedSVM, svmv1alpha1.MigrationRunning, migrationRunningStatusReason),
+			metav1.UpdateOptions{},
+		)
+	if err != nil {
+		return err
+	}
+
+	// ToDo: implement a mechanism to resume migration from the last migrated resource in case of a failure
+	// process storage migration
+	gvk, err := svmc.restMapper.KindFor(gvr)
+	if err != nil {
+		return err
+	}
+
+	for _, gvrKey := range svmc.gcCollector.GetMonitorStore(gvr).ListKeys() {
+		namespace, name, err := cache.SplitMetaNamespaceKey(gvrKey)
+		if err != nil {
+			return err
+		}
+
+		typeMeta := metav1.TypeMeta{}
+		typeMeta.APIVersion, typeMeta.Kind = gvk.ToAPIVersionAndKind()
+		data, err := json.Marshal(typeMeta)
```

Comment:
This is the same for every iteration of this loop, right?  Can this be extracted out of the loop?